common/installshield_cab.cpp

```diff
@@ -48,114 +48,177 @@
 #include "common/memstream.h"
+#include "common/substream.h"
+#include "common/ptr.h"
 #include "common/zlib.h"
 
 namespace Common {
 
+namespace {
+
 class InstallShieldCabinet : public Archive {
 public:
-	InstallShieldCabinet(SeekableReadStream *stream, DisposeAfterUse::Flag disposeAfterUse);
-	~InstallShieldCabinet();
+	InstallShieldCabinet();
+
+	bool open(const String &baseName);
+	void close();
 
 	// Archive API implementation
-	bool hasFile(const String &name) const;
-	int listMembers(ArchiveMemberList &list) const;
-	const ArchiveMemberPtr getMember(const String &name) const;
-	SeekableReadStream *createReadStreamForMember(const String &name) const;
+	virtual bool hasFile(const String &name) const;
+	virtual int listMembers(ArchiveMemberList &list) const;
+	virtual const ArchiveMemberPtr getMember(const String &name) const;
+	virtual SeekableReadStream *createReadStreamForMember(const String &name) const;
 
 private:
 	struct FileEntry {
 		uint32 uncompressedSize;
 		uint32 compressedSize;
 		uint32 offset;
 		uint16 flags;
+		uint16 volume;
 	};
 
+	int _version;
 	typedef HashMap<String, FileEntry, IgnoreCase_Hash, IgnoreCase_EqualTo> FileMap;
 	FileMap _map;
-	Common::SeekableReadStream *_stream;
-	DisposeAfterUse::Flag _disposeAfterUse;
-};
+	String _baseName;
 
-InstallShieldCabinet::~InstallShieldCabinet() {
-	_map.clear();
+	String getHeaderName() const;
+	String getVolumeName(uint volume) const;
+};
 
-	if (_disposeAfterUse == DisposeAfterUse::YES)
-		delete _stream;
+InstallShieldCabinet::InstallShieldCabinet() : _version(0) {
 }
 
-InstallShieldCabinet::InstallShieldCabinet(SeekableReadStream *stream, DisposeAfterUse::Flag disposeAfterUse) : _stream(stream), _disposeAfterUse(disposeAfterUse) {
-	// Note that we only support a limited subset of cabinet files
-	// Only single cabinet files and ones without data shared between
-	// cabinets.
+bool InstallShieldCabinet::open(const String &baseName) {
+	// Store the base name so we can generate file names
+	_baseName = baseName;
+
+	// Try to find the file
+	SeekableReadStream *header = SearchMan.createReadStreamForMember(getHeaderName());
+	if (!header)
+		header = SearchMan.createReadStreamForMember(getVolumeName(1));
+
+	if (!header) {
+		close();
+		return false;
+	}
+
+	// Note that we only support v5 and v6 cabinet files
 
 	// Check for the magic uint32
-	uint32 magic = _stream->readUint32LE();
+	uint32 magic = header->readUint32LE();
 	if (magic != 0x28635349) {
 		warning("InstallShieldCabinet::InstallShieldCabinet(): Magic ID doesn't match: expecting %x but got %x", 0x28635349, magic);
-		return;
+		close();
+		return false;
 	}
 
-	uint32 version = _stream->readUint32LE();
-
-	if (version != 0x01000004) {
+	uint32 version = header->readUint32LE();
+
+	switch(version) {
+	case 0x01000004:
+		_version = 5;
+		break;
+	case 0x0100600c:
+		_version = 6;
+		break;
+	case 0x020004B0: // Found in some Russian variants of Nancy Drew games. Possibly a malformed header
```

Comment:
This is actually InstallShield version 12, not a malformed header. Check https://github.com/twogood/unshield/blob/master/lib/libunshield.c and https://www.onlinegdb.com/mXzjHjXAt for a small program that extracts the version, based on that repo (note that magic bytes 0x01000004 are actually transformed into version 0 using that algorithm).

Please change this